src/Stratis.Bitcoin/BlockPulling2/BlockPuller.cs

```diff
@@ -0,0 +1,1005 @@
+            {
+                try
+                {
+                    await this.processQueuesSignal.WaitAsync(this.cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    this.logger.LogTrace("(-)[CANCELLED]");
+                    return;
+                }
+
+                await this.AssignDownloadJobsAsync().ConfigureAwait(false);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Loop that continuously checks if peers failed to deliver important blocks in given time and penalizes them if they did.</summary>
+        private async Task StallingLoopAsync()
+        {
+            this.logger.LogTrace("()");
+
+            while (!this.cancellationSource.IsCancellationRequested)
+            {
+                try
+                {
+                    await Task.Delay(StallingLoopIntervalMs, this.cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    this.logger.LogTrace("(-)[CANCELLED]");
+                    return;
+                }
+                
+                this.CheckStalling();
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+        
+        /// <summary>Assigns downloads from <see cref="reassignedJobsQueue"/> and <see cref="downloadJobsQueue"/> to the peers that are capable of delivering blocks.</summary>
+        private async Task AssignDownloadJobsAsync()
+        {
+            this.logger.LogTrace("()");
+
+            var failedHashes = new List<uint256>();
+            var newAssignments = new List<AssignedDownload>();
+
+            lock (this.queueLock)
+            {
+                // First process reassign queue ignoring slots limitations.
+                this.ProcessQueueLocked(this.reassignedJobsQueue, newAssignments, failedHashes);
+
+                // Process regular queue.
+                int emptySlots;
+                lock (this.assignedLock)
+                {
+                    emptySlots = this.maxBlocksBeingDownloaded - this.assignedDownloadsByHash.Count;
+                }
+
+                this.logger.LogTrace("There are {0} empty slots.", emptySlots);
+
+                if (emptySlots > this.maxBlocksBeingDownloaded * MinEmptySlotsPercentageToStartProcessingTheQueue)
+                    this.ProcessQueueLocked(this.downloadJobsQueue, newAssignments, failedHashes, emptySlots);
+                
+                this.processQueuesSignal.Reset();
+            }
+            
+            // Call callbacks with null since puller failed to deliver requested blocks.
+            this.logger.LogTrace("{0} jobs partially or fully failed.", failedHashes.Count);
+            foreach (uint256 failedJob in failedHashes)
+                this.OnDownloadedCallback(failedJob, null);
```

Comment:
What does failed mean? if no peer was found that can download this hash?

Reply:
Yes. There are no peers claiming requested hash.